Инкремент 9

Fairness2 opened the pull request:
Добавлено сжатие GZIP для сервера и клиента. Добавил хранилище с записью в файл. API переведено на JSON на сервере и агенте. Добавлено логирование на серере с помощью ZAP

Fairness2 committed:
iter6 | Добавлена система логирования сервера

В представленных изменениях была добавлена система логирования, с возможностью настройки уровня логирования. Также были созданы middleware для логирования входящих запросов и исходящих ответов.

Список измененных файлов:
cmd/agent/main.go
cmd/server/main.go
cmd/skill/flags.go
internal/logger/skill/logger.go
internal/logger/logger.go
go.mod
cmd/server/config/config.go
internal/logger/middleware.go
go.sum
cmd/agent/config/env.go
cmd/agent/collector/sender/sender_test.go
cmd/skill/main.go

Fairness2 committed:
iter7 | Улучшена обработка запросов метрик, изменены тесты

Было добавлено разделение обработчиков запросов на два типа: для взаимодействия через URL и через JSON тело запроса. Также были произведены некоторые изменения в тестах.

Список измененных файлов:
cmd/server/handlers/handlemetric/handlemetric.go
cmd/server/handlers/getmetric/handler_test.go
cmd/server/handlers/handlemetric/jsonhandler_test.go
cmd/server/handlers/handlemetric/handlemetric_test.go
cmd/agent/collector/sender/sender_test.go
cmd/skill/main.go
cmd/server/handlers/handlemetric/urlhandler.go
cmd/server/handlers/getmetrics/handler.go
cmd/server/handlers/getmetric/urlhandler.go
cmd/agent/main.go
internal/helpers/helpers.go
cmd/server/main.go
internal/helpers/helpers_test.go
internal/models/skill/models.go
cmd/skill/main_test.go
cmd/server/handlers/handlemetric/errors.go
cmd/server/config/config.go
cmd/server/handlers/handlemetric/service.go
internal/metrics/Metrics.go

Fairness2 committed:
iter7.1 | Дополнили функционал и улучшили тесты модулей обработки и отправки метрик

Импортирована middleware для удаления лишних слэшей в URL. Вынесен код функции сбора stats в отдельную функцию collect(). В функции отправки метрик использован новый метод GetRaw() для получения данных метрики в изначальном типе. В обработке JSON тела запроса теперь устанавливается статус ответа 200, если выполнено без ошибок.

Список измененных файлов:
cmd/server/main.go
cmd/agent/collector/collection/initializer.go
cmd/agent/collector/sender/sender_test.go
cmd/agent/collector/sender/sender.go
cmd/server/handlers/handlemetric/jsonhandler.go
cmd/server/handlers/getmetric/jsonhandler.go

Fairness2 committed:
iter7.2 | Изменен механизм начальной отправки метрик

Был изменен механизм отправки метрик в начале работы программы: теперь эта операция происходит через одну секунду после старта, а не сразу после. Это позволяет собрать более полное количество метрик для первой отправки.

Список измененных файлов:
cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter8 | Добавлена поддержка сжатия GZIP в клиенте и сервере

Была проведена работа по добавлению поддержки сжатия в формате GZIP для входящих и исходящих запросов в клиенте и сервере. На стороне сервера были добавлены middleware для обработки сжатых запросов и ответов. На стороне клиента при отправке данных на сервер теперь используется сжатие.

Список измененных файлов:
cmd/agent/main.go,
cmd/server/main.go,
internal/helpers/compress/compress.go,
cmd/skill/main_test.go,
internal/middlewares/middlewares.go,
cmd/agent/collector/sender/sender.go,
cmd/server/config/parse.go,
cmd/skill/main.go,
cmd/skill/gzip.go.

Fairness2 committed:
iter8.1 | Изменено имя типа HttpWriter на HTTPWriter и соответствующие функции

Изменённые файлы:
internal/helpers/compress/compress.go

Fairness2 committed:
iter9 | Добавлено хранилище метрик с записью в файл

Было добавлено хранилище метрик с возможностью записи данных в файл, с выбором синхронного и асинхронного режимов хранения. Конфигурация сервера расширена параметрами для настройки этого хранилища.

Измененные файлы:
cmd/server/main.go,
internal/metrics/MemStorage.go,
internal/contextkeys/keys.go,
cmd/server/config/config.go,
internal/metrics/fileworker/writer.go,
cmd/server/config/parse.go,
internal/metrics/FileStorage.go

Fairness2 committed:
iter9.1 | Переработано управление хранилищем метрик и логгером

В коде сервера были осуществлены изменения, которые сделали управление хранилищем метрик и логгером более гибким благодаря введению функции инициализации. Также исправлено восстановление из файла при запуске в случае, если файл не существует или пуст.

Измененные файлы:
cmd/server/main.go,
cmd/server/config/config.go,
internal/metrics/FileStorage.go

Fairness2 committed:
iter9.2 | Добавлен обработчик сигналов для остановки промежуточного хранилища и логгера

Внесены изменения в основной код сервера для управления закрытием промежуточного хранилища и логгера: добавлен обработчик системных сигналов для гибкой остановки. Также изменено значение пути хранения метрик по умолчанию в конфигурационном файле.

Измененные файлы:
cmd/server/main.go,
cmd/server/config/config.go.

Fairness2 committed:
iter9.3 | Добавлена гибкая остановка сервера и улучшено закрытие хранилища

Была добавлена библиотека manners для гибкой остановки сервера на Go, позволяющая ему дождаться окончания всех текущих запросов. Был переработан механизм закрытия хранилища, обеспечивающий синхронизацию данных перед его закрытием.

Измененные файлы:
cmd/server/main.go
go.mod
go.sum
internal/metrics/FileStorage.go

Fairness2 committed:
iter9.4 | Добавлен контекст для корректной остановки и синхронизации сервера

В cmd/server/main.go внесены изменения для того, чтобы добавить контекст в запуск сервера. Это сделано для того, чтобы корректно останавливать и синхронизировать работу сервера и горутин, связанных с ним.

Измененный файл:
cmd/server/main.go

Fairness2 committed:
iter9.5 | Обновили сервер для использования стандартной библиотеки HTTP

Произведена замена библиотеки manners на стандартную библиотеку net/http для работы сервера. Введен контекст для корректного завершения сервера и добавлена синхронизация с помощью WaitGroup.

Измененные файлы:
- cmd/server/main.go
- go.mod

Fairness2 committed:
iter9.6 | Перенесли объявления переменных в константы

Измененные файлы:
- cmd/server/config/config.go
- cmd/agent/config/env.go

vokinneberg commented on cmd/agent/collector/collection/initializer.go:
```diff
@@ -27,18 +27,23 @@ func NewCollection() *Type {
 func CollectProcess(ctx context.Context) {
 	log.Printf("Collect metrics process starts. Period is %d mseconds\n", config.Params.PollInterval)
+	// Делаем первый сбор метрик сразу же
+	collect()
```
👍 за переработку этой функции.

vokinneberg commented on cmd/agent/collector/collection/initializer.go:
```diff
@@ -27,18 +27,23 @@ func NewCollection() *Type {
 }
+
+// collect reads memory stats and calls Collection.Collect to populate the collection with system metrics.
+func collect() {
```
👍 за реализацию отдельной функции сбора метрик.

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -29,50 +32,68 @@ func New(mCollection *collection.Type) *Client {
 func (c *Client) PeriodicSender(ctx context.Context) {
 	log.Println("Starting periodic sender")
+	// В первый раз отправляем метрики через одну секунду, чтобы дать собрать метрики
+	<-time.After(1 * time.Second)
```
Мне не кажется такой подход достаточно надежным. Что произойдет, если по какой-то причине метрики не будут собраны за 1 секунду? Думаю тут стоит подумать о том, как реализовать первую отправку метрик гарантировано после того, как они будут собраны?

Fairness2 replied:
Теперь первая отправка будет по таймеру

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -29,50 +32,68 @@ func New(mCollection *collection.Type) *Client {
+	}
 	go func() {
-		err := c.sendMetric(metrics.TypeCounter, "PollCount", pCnt)
+		err := c.sendMetric(body)
```
Видимо я не заметил этого поведения во время проверки предидущего инкремента, но тут я вижу потенциальную проблему с тем, что счетчик `c.metricsCollection.ResetCounter()` будет сброшен до того, как метрики будут отправлены. Возможно вызов горутины тут лишний и синхронная отправка метрик позволит упростить логику.

Fairness2 replied:
Возможно данные не придут последовательно и каунтер может прийти раньше некоторых метрик. В данном случае данные метрики не изменятся при отправке в горутинах, так как передаются по значению, а не по ссылке 🤔

Fairness2 replied:
В третьем спринте отправка будет изменена на метод, где все метрики отправляются разом, может там и переделать?

Fairness2 replied:
В данном случае отправляемые данные не будут изменены, так как будут переданы по значению. В третьем спринте будет переделано на отправку пакетом. Пока сделал группу ожидания, чтобы сбросить каунтер только после отправки всех метрик

vokinneberg replied:
ОК, давай оставим вопрос на 3 спринт 👍

vokinneberg commented on cmd/agent/collector/sender/sender.go:
```diff
@@ -82,34 +103,63 @@ func (c *Client) sendMetrics() {
+		return nil, err
+	}
+
+	// Создаём буфер, в который запишем сжатое тело
```
Если хочется оптимизировать выделение памяти можно посмотреть в сторону `sync.Pool`. 

```go
 var gzipWriterPool = sync.Pool{
	New: func() interface{} {
		writer, _ := gzip.NewWriterLevel(nil, gzip.BestSpeed)
		return writer
	},
}
```

И вместо создания буфера получать инстанс из пула:

```go
        var buf bytes.Buffer
	writer := gzipWriterPool.Get().(*gzip.Writer)
	defer gzipWriterPool.Put(writer)
```

Fairness2 replied:
Переделал на Pool

Fairness2 replied:
Сделал группу врайтеров для отправки

vokinneberg commented on cmd/agent/config/env.go:
```diff
@@ -18,23 +18,25 @@ type CliConfig struct {
 
-// DefaultPollInterval Интервал между сборкой данных по умолчанию
-var DefaultPollInterval = 2 * time.Second
+const (
```
Рекомендую поднять `const` блок на верх файла сразу после определения импортов, это более ковенциональное место для определения констант.

Fairness2 replied:
Перенёс всё выше

vokinneberg commented on cmd/agent/main.go:
```diff
@@ -17,7 +20,7 @@ func main() {
 		log.Fatal(err)
 	}
-	config.SetGlobalConfig(cnf)
+	config.Params = cnf
```
Я по-моему об этом уже писал - не понятно зачем нам ограничивать работу клиента 30 секундами. Мне кажется тут нужно создать не таймаут контекст, а контекст с отменой:

```go
    ctx, cancel := context.WithCancel(context.Background())
```

Fairness2 replied:
УБрал, вроде теперь тесты проходят)

vokinneberg commented on internal/logger/middleware.go:
```diff
@@ -0,0 +1,72 @@
+
+// LogRequests мидлеваре, которое регистрирует данные запроса
+// Функция регистрирует метод, путь и продолжительность каждого запроса
+func LogRequests(next http.Handler) http.Handler {
```
Все-таки по условию задачи логирование запросов и ответов должно быть реализовано в виде одного middleware хендлера, который сначала логирует запрос, а потом ответ. Мне кажется, это принципиально с точки зрения понимания работы хендлера.

Fairness2 replied:
Объединил LogRequests и LogResponse

vokinneberg replied:
Здорово, спасибо! 👍

vokinneberg commented on internal/logger/middleware.go:
```diff
@@ -0,0 +1,72 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		// Регистрируем завершающую функцию, чтобы залогировать в любом случае
+		defer func() {
```
Здесь логирование запроса произойдет после обработки запроса, мне кажется это не совсем то чего мы хотим добиться?

Fairness2 replied:
Данные о времени выполнения запроса, статусе и размере ответа можно записать только после выполнения

vokinneberg replied:
В целом оба подхода валидны. Я имел ввиду то, что если мы залогируем данные запроса до того, как он будет обработан, то в случае сбоя у нас останется лог запроса хотя бы.

Fairness2 replied:
В третьем спринте попробую переделать

vokinneberg commented on internal/logger/skill/logger.go:
```diff
@@ -0,0 +1,43 @@
+var Log *zap.Logger = zap.NewNop()
+
+// Initialize инициализирует синглтон логера с необходимым уровнем логирования.
+func Initialize(level string) error {
```
Мне кажется этот код как минимум дублирует код в файле `internal/logger/logger.go`. Можем ли мы сделать общую реализацию инициализации логгера?

Fairness2 replied:
Сделал общую реализацию. Но вообще, всё что связано с skill основано на примерах, что написаны в конце глав, где рассказывается пример создания навыка для Алисы. Было написано, что можно их делать и на основной проект не повлияет)

vokinneberg replied:
Я стараюсь комментировать весь код. На проект это не влияет, но мне кажется даже в опциональном коде следует следовать лучшим практикам, так сказать набивать руку 😊

vokinneberg commented on cmd/server/config/parse.go:
```diff
@@ -25,18 +28,52 @@ func Parse() (*CliConfig, error) {
-	envAddr := os.Getenv("ADDRESS")
-	if envAddr != "" {
-		params.Address = envAddr
+	cnf := struct {
```
Мы уже определили такую же структуру в файле `cmd/server/config/config.go`. Можем ее тут переиспользовать?

Fairness2 replied:
Не получается, так как при использовании типа time.Duration библиотека предполагает, что время в секундах будет передано в виде 23s, то есть с мерой времени, а по постановке задачи требуется прописывать 23 просто, поэтому в этой структуре используется int. Если бы не это требование, то можно было бы использовать структуру из config.go

vokinneberg replied:
А что нам мешает там использовать тип int? Мне кажется, что если будет единобразно, будет только лучше?

Fairness2 replied:
Подумаю как это решить

vokinneberg commented on cmd/server/config/parse.go:
```diff
@@ -25,18 +28,52 @@ func Parse() (*CliConfig, error) {
+	if cnf.FileStorage != "" {
+		params.FileStorage = cnf.FileStorage
+	}
+	if _, ok := os.LookupEnv("RESTORE"); ok {
+		params.Restore = cnf.Restore
+	}
```
Не совсем понятна логика иницализации? Зачем поверять `os.LookupEnv("RESTORE")`, но при этом устанавливать `params.Restore = cnf.Restore`? Не проще ли просто установить значением сразу равным `cnf.Restore`?
 
```suggestion
	params.Restore = cnf.Restore
```

Fairness2 replied:
Тут я так сделал, чтобы мы могли поменять значение по умолчанию в константе, там же где все остальные значения по умолчанию. Иначе, при изменении значения по умолчанию, если не указывать, RESTORE, то будет записано значение по умолчанию из структуры для парсинга. Так как у нас нет пустого значения bool, то проверить установлено или нет RESTORE, не получится

vokinneberg commented on cmd/server/main.go:
```diff
@@ -17,31 +27,146 @@ func main() {
+type next func()
+
+// InitStore устанавливаем глобальное хранилище метрик.
+func InitStore(n next, ctx context.Context, wg *sync.WaitGroup) {
```
По конвенции Go, контекст передается как первый параметр функции.

```suggestion
func InitStore(ctx context.Context, n next,  wg *sync.WaitGroup) {
```

Fairness2 replied:
Перенёс и упростил функции

vokinneberg commented on cmd/server/main.go:
```diff
@@ -17,31 +27,146 @@ func main() {
+		cancel()
+	}()
+
+	InitLogger(func() {
```
Не совсем понятно, какую задачу решает такой последовательный вызов функций?  По-моему код стало сложнее понимать, при этом просто последовательный вызов функций решил ту же задачу без проблем.

Fairness2 replied:
Идея была в том, чтобы вынести инициализацию компонентов в отдельные функции, Переделал по другому, но теперь дефер функция по закрытию хранилища в другой функции

vokinneberg commented on cmd/server/main.go:
```diff
@@ -17,31 +27,146 @@ func main() {
-	metrics.MeStore = metrics.NewMemStorage()
+	wg := sync.WaitGroup{} // Группа для синхронизации
+	defer func() {
+		wg.Wait() // Ожидаем завершения всех горутин перед завершением программы
```
Я бы не рекомендовал вызывать `wg.Wait()` в `defer` блоке, так как это неявно - логика же `wg.Wait()` предполагает явную блокировку и ожидание на ней.

Fairness2 replied:
Вынес всё что использует группу в отдельную функцию, а в мэйне оставил только ожидание

Fairness2 replied:
Переделал этот момент, вынес код в новую функцию, которая вызывается и группа ожидает в конце вызывающей функции теперь

vokinneberg commented on cmd/server/main.go:
```diff
@@ -17,31 +27,146 @@ func main() {
+		if !store.SyncMode {
+			wg.Add(1)
+			go func() {
+				store.Sync(ctx)
+				defer wg.Done()
```
`defer` должен вызываться как можно раньше после входа в функцию. Если краш произойдет в `store.Sync(ctx)` defer ни когда не выполнится.

```suggestion
				defer wg.Done()
				store.Sync(ctx)
```

Fairness2 replied:
Поправил

Fairness2 replied:
Перенёс

vokinneberg commented on cmd/skill/main_test.go:
```diff
@@ -62,3 +86,71 @@ func TestWebhook(t *testing.T) {
+		defer resp.Body.Close()
+
+		b, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
```
Было бы здорово добавить информативные сообщения о том, что пошло не так и что ожидалось.

Fairness2 replied:
Добавил информационные сообщения

vokinneberg commented on cmd/skill/main.go:
```diff
@@ -49,11 +61,77 @@ func webhook(response http.ResponseWriter, request *http.Request) {
+	logger.Log.Debug("sending HTTP 200 response")
+}
+
+func gzipMiddleware(h http.Handler) http.Handler {
```
Можем ли мы использовать те же middleware, которые используем для сервера?

Fairness2 replied:
Переделал на использование одних мидлварей

vokinneberg commented on internal/metrics/FileStorage.go:
```diff
@@ -0,0 +1,124 @@
+type Writer interface {
+	Write(v any) error
+	Close() error
+}
```
Здесь можно попробовать использовать комбинацию интерфейсов и исользовать уже существующий интерфейс из пакета `io`.

```suggestion
type Writer interface {
	Write(v any) error
	io.Closer
}
```

Fairness2 replied:
Добавил интерфейс io.Closer

vokinneberg reviewed (changes requested):
Поздравляю с первым ревью второго спринта! 🤗  В этом спринте мы охватили большой и важный материал, а наше приложение усложнилось. 💪 Ты справился с реализацией и реализовал все в соотвествии с постановкой задачи. Отличная работа! 🚀 Тем не менее мне хотелось бы добавить несколько комментариев относительно твоего кода и сделать несколько предложений по улучшению. 🤓 Так же предлагаю запланировать встречу 1 на 1, где мы могли бы обсудить review 🤝

**Сильные стороны твоего PR:**

1. Код функционален и корректен! 🚀
2. Ты реализовал тесты новых сценариев и дополнил существущие. 👍
3. Ты реализовал graceful shutdown для клиента и сервера и разобрался с синхронизацией горутин и работой с каналами и Wait группами. Очень круто 👍

**На что я бы рекомендовал обратить внимание и что можно улучшить:**

1. Стоит поправить 30 секундный таймаут в клиенте - это основаная проблема, которую я вижу в твоем коде.
2. Мне показалось, что реализация запуска сервера и graceful shutdown  реализованы немного сложно и эту часть можно было бы упростить. Не стоит злоупотреблять вызовом горутин и стоит помнить о том, что конкурентный код  сложнее отлаживать и понимать.

Fairness2 committed:
iter9 | Удалены навыки gzip и логер из skill, реорганизована обработка отправки метрик агента, улучшено ведение журналов и промежуточное программное обеспечение.

Удалены gzip и logger из skill, вместо них используются общие компоненты. Изменена обработка бработка отправки метрик агента, теперь используется пул врайтеров, улучшена логика ведения логов и работы промежуточных слоев. Обновлены соответствующие тесты и конфигурации.

Измененные файлы:
- cmd/skill/gzip.go
- internal/logger/skill/logger.go
- cmd/agent/main.go
- internal/helpers/helpers.go
- cmd/server/main.go
- cmd/skill/main_test.go
- cmd/server/config/config.go
- internal/middlewares/middlewares.go
- cmd/agent/config/env.go
- cmd/skill/main.go
- internal/metrics/FileStorage.go
- cmd/server/handlers/handlemetric/jsonhandler.go
- cmd/server/handlers/getmetric/jsonhandler.go
- internal/logger/logger.go
- internal/logger/middleware.go

Fairness2 committed:
iter9 | Заменены time.After на time.NewTicker и добавлены ticker.Stop в циклических функциях

Это изменение устраняет потенциальную утечку ресурсов и улучшает управление временем, используя time.NewTicker вместо time.After для циклических операций. ticker.Stop был добавлен для корректной остановки таймера при завершении контекста.

Измененные файлы:
- cmd/agent/collector/collection/initializer.go
- cmd/agent/collector/sender/sender.go
- internal/metrics/FileStorage.go

Fairness2 committed:
iter9 | Добавлен контроль завершения потоков при отправке метрик

Использован WaitGroup для ожидания завершения всех горутин, отправляющих метрики. Это гарантирует сброс счетчика только после отправки всех данных.

Изменённые файлы:
cmd/agent/collector/sender/sender.go

Fairness2 committed:
iter9 | Добавил сообщения об ошибках в require.NoError и require.Equal

Улучшил тесты для вывода более информативных сообщений об ошибках. Теперь в require.NoError и require.Equal передаются сообщения, что позволяет легче диагностировать проблемы.

Измененные файлы:
- cmd/skill/main_test.go

vokinneberg reviewed (approved):
Поздравляю с успешным ревью второго спринта! 🤗 Ты реализовал многие из моих предложений. Здорово! Мне кажется, удалось упростить некоторые моменты, и код стал понятнее. У нас остался открытый вопрос относительно сброса счетчика, я предлагаю обсудить его на 3 спринте. 🤝
